### PR TITLE
🦩 argocd - fix CRD race condition 🦩

### PR DESCRIPTION
**Pull request by eformat**
this fixes the race condition when installing argocd into an empty cluster using helm lifecycle hooks, waiting for CRD to be made available by subscription/OLM

the ramifications are that 'helm template' does not recognise these hooks, this will affect https://github.com/rht-labs/ubiquitous-journey usage e.g. bootstrap would look like this

```
# bootstrap to install argocd and create projects
helm upgrade --install bootstrap -f bootstrap/values-bootstrap.yaml bootstrap --create-namespace --namespace labs-bootstrap
# uninstall bootstrap chart
helm delete bootstrap --namespace labs-bootstrap --debug
```

### CODE REVIEW

**Comment by garethahealy** [comment]
@eformat ; just for my own understanding, how does helm know to wait for the `wait-for-crd` pod to have run? was looking at the other bits but couldn't work it out by reading the YAML.

**Comment by eformat** [comment]
> @eformat ; just for my own understanding, how does helm know to wait for the `cluster-check` pod to have run? was looking at the other bits but couldn't work it out by reading the YAML.

The pattern looks like this:

(1) for operator subscriptions

normal yaml/helm deploy in the templates/folder

(2) then wait for CRD's pod 

this uses a hook, with priority that makes it run before other post-install hooks, and deletes itself once OK

  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-weight": "-1"
    "helm.sh/hook-delete-policy": hook-succeeded,hook-failed

(3) then for anything that requires a CRD e.g. a CR, deploy using hook, but dont delete after success, needs to run after crd check pod in (2)

  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-weight": "25"

File: charts/argocd-operator/templates/wait-for-crd.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "helm.sh/hook": post-install,post-upgrade
+    "helm.sh/hook-weight": "-1"
+    "helm.sh/hook-delete-policy": hook-succeeded
+  namespace: {{ .Values.namespace }}
+spec:
+  containers:
+  - name: crd-check 
+    image: quay.io/openshift/origin-cli:4.6
+    imagePullPolicy: IfNotPresent
+    command: ['sh', '-c', 'while [ true ]; do oc get crd argocds.argoproj.io; if [ $? -eq 0 ]; then break; fi ; sleep 5s; done']
```
[the remark this answers is not in the document]
Comment (eformat) [reply charts/argocd-operator/templates/wait-for-crd.yaml:16]: yeah .. for completeness i can add no probs ...

File: charts/argocd-operator/templates/delete-csv-hook-rbac.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: csv-deleters
+  namespace: {{ .Values.namespace }}
+  annotations:
+    "helm.sh/hook": pre-delete
+    "helm.sh/hook-weight": "1"
+    "helm.sh/hook-delete-policy": hook-succeeded
```
[the remark this answers is not in the document]
Comment (eformat) [reply charts/argocd-operator/templates/delete-csv-hook-rbac.yaml:35]: good call dude ...

**Comment by eformat** [comment]
@springdo @tylerauerbeck i have also added support for an **ignoreHelmHooks** flag to this chart. see:

https://github.com/redhat-cop/helm-charts/pull/120/commits/f90f989d3296eff454c217c2200c1d9f6835be52

this is for ArgoCD support when you want to manage the chart as a dependency in argocd e.g. UJ bootstrap - this is something that currently works for "helm template" .. but because of the way hooks map in argocd (https://argo-cd.readthedocs.io/en/stable/user-guide/helm/#helm-hooks) you get the wrong behavior if you use helm to first install the chart.

so, the flow looks like this (my testing so far using a petbattle repository)
```
git clone https://github.com/petbattle/pb-ci-cd.git
cd ubiquitous-journey

helm upgrade --install bootstrap \
  -f bootstrap/values-bootstrap.yaml \
  bootstrap --create-namespace --namespace labs-bootstrap

argocd login $(oc get route argocd-server --template='{{ .spec.host }}' -n labs-ci-cd):443 --sso --insecure

argocd app create bootstrap-journey \
  --dest-namespace labs-bootstrap \
  --dest-server https://kubernetes.default.svc \
  --repo https://github.com/petbattle/pb-ci-cd.git \
  --revision main \
  --sync-policy automated \
  --path "ubiquitous-journey/bootstrap" \
  --helm-set argocd-operator.ignoreHelmHooks=true \
  --values "values-bootstrap.yaml" 
```

Screenshot from 2021-02-16 12-32-32

**Comment by springdo** [comment]
LGTM 🪂

**Pull request merged**